### PR TITLE
Initial Pipelines Code

### PR DESCRIPTION
**Pull request by Im5tu**
All commits will be squashed on merge. See the readme for usage :)

### CODE REVIEW

File: src/Conduit/IPipelineManager.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    /// Manages the creation of pipelines for multiple types
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
+    public interface IPipelineManager<T>
+    {
+        /// <summary>
+        /// Returns the pipeline with the specified name
+        /// </summary>
+        /// <param name="name">The name of the pipeline</param>
+        /// <returns>The asynchronous pipeline</returns>
```
Comment (comdw) [inline src/Conduit/IPipelineManager.cs]:
```suggestion
        /// <returns>The pipeline</returns>
```

File: src/Conduit/IPipelinePhase.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Conduit
+{
+    /// <summary>
+    /// Defines a phase in the pipeline in which processing can occur
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/IPipelinePhase.cs]:
```suggestion
    /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/IAsyncPipelinePhase.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+
+namespace Conduit
+{
+    /// <summary>
+    /// Defines a phase in the pipeline in which processing can occur asynchronously
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/IAsyncPipelinePhase.cs]:
```suggestion
    /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/ConduitServiceExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+{
+    /// <summary>
+    /// Extensions for the Conduit library
+    /// </summary>
+    public static class ConduitServiceExtensions
+    {
+        /// <summary>
+        /// Adds the default pipeline for a given type
+        /// </summary>
+        /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/ConduitServiceExtensions.cs]:
```suggestion
        /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/ConduitServiceExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+                throw new ArgumentNullException(nameof(builderAction));
+            }
+
+            return services.BuildPipeline(builderAction).AddPipelineServices<T>();
+        }
+
+        /// <summary>
+        /// Adds a pipeline with the specified name which can be retrieved from IPipelineManger&lt;T&gt;.Get(name)
+        /// </summary>
+        /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/ConduitServiceExtensions.cs]:
```suggestion
        /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/ConduitServiceExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+                throw new ArgumentNullException(nameof(builderAction));
+            }
+
+            return services.BuildPipeline(builderAction, name).AddPipelineServices<T>();
+        }
+
+        /// <summary>
+        /// Adds the default pipeline for a given type
+        /// </summary>
+        /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/ConduitServiceExtensions.cs:65]:
```suggestion
        /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/ConduitServiceExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                throw new ArgumentNullException(nameof(builderAction));
+            }
+
+            return services.BuildPipeline(builderAction).AddPipelineServices<T>();
+        }
+
+        /// <summary>
+        /// Adds a pipeline with the specified name which can be retrieved from IAsyncPipelineManger&lt;T&gt;.Get(name)
+        /// </summary>
+        /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/ConduitServiceExtensions.cs:87]:
```suggestion
        /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/IAsyncPipelineBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Conduit
+{
+    /// <summary>
+    /// Builder interface for constructing a pipeline of the specified type
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/IAsyncPipelineBuilder.cs:11]:
```suggestion
    /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/IAsyncPipeline.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+
+namespace Conduit
+{
+    /// <summary>
+    /// Represents an asynchronous execution pipeline
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/IAsyncPipeline.cs:9]:
```suggestion
    /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/IAsyncPipelineManager.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Conduit
+{
+    /// <summary>
+    /// Manages the creation of pipelines for multiple types
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/IAsyncPipelineManager.cs]:
```suggestion
    /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/IPipeline.cs [hunk 11]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Conduit
+{
+    /// <summary>
+    /// Represents an execution pipeline
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/IPipeline.cs:6]:
```suggestion
    /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/IPipelineBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using System;
+
+namespace Conduit
+{
+    /// <summary>
+    /// Builder interface for constructing a pipeline of the specified type
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/IPipelineBuilder.cs:9]:
```suggestion
    /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: src/Conduit/IPipelineManager.cs [hunk 13]
```diff
@@ -0,0 +1,16 @@
+﻿namespace Conduit
+{
+    /// <summary>
+    /// Manages the creation of pipelines for multiple types
+    /// </summary>
+    /// <typeparam name="T">The that is passed through the pipeline</typeparam>
```
Comment (comdw) [inline src/Conduit/IPipelineManager.cs:6]:
```suggestion
    /// <typeparam name="T">The type that is passed through the pipeline</typeparam>
```

File: README.md [hunk 14]
```diff
@@ -1,2 +1,109 @@
...
+})
+```
+
+As part of the default pipeline builder, you have access to all of the same stuff that you would when defining a named pipeline.
+
+Once this is registered, you can access the pipeline in your code by injecting `IPipeline<T>`/`IAsyncPipeline<T>`.
+
+_**Note**: You can have different default pipelines for your synchronos and asynchronos code paths._
+
+### Conditional execution of pipline phases
```
Comment (comdw) [inline README.md]:
```suggestion
### Conditional execution of pipeline phases
```

File: src/Conduit/PipelineBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    internal sealed class PipelineBuilder<T> : IPipelineBuilder<T>
+    {
+        private readonly IServiceCollection _services;
+        private readonly string _name;
+        private readonly List<Func<IServiceProvider, T, T>> _phases = new List<Func<IServiceProvider, T, T>>();
+
+        public PipelineBuilder(IServiceCollection services, string name)
+        {
+            _services = services ?? throw new ArgumentNullException(nameof(services));
+            _name = name ?? throw new ArgumentNullException(nameof(name));
```
Comment (comdw) [inline src/Conduit/PipelineBuilder.cs:16]: is empty string ok?

File: src/Conduit/PipelineManager.cs [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+            _serviceScopeFactory = serviceScopeFactory ?? throw new ArgumentNullException(nameof(serviceScopeFactory));
+        }
+
+        public IPipeline<T> Get(string name)
+        {
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                name = "_default";
```
Comment (comdw) [inline src/Conduit/PipelineManager.cs]: Isn't the default pipeline called Constants.DefaultPipeline?
```suggestion
                name = Constants.DefaultPipeline;
```

File: src/Conduit/PipelineManager.cs [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+            }
+
+            var options = _options.Get(name) ?? throw new Exception("Cant find it");
```
Comment (comdw) [inline src/Conduit/PipelineManager.cs]:
```suggestion
            var options = _options.Get(name) ?? throw new ArgumentException($"Can't find named pipeline ${name}", name);
```

File: src/Conduit/AsyncPipelineManager.cs [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+        }
+
+        public IAsyncPipeline<T> Get(string name)
+        {
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                name = Constants.DefaultPipeline;
+            }
+
+            var options = _options.Get(name) ?? throw new Exception("Cant find it");
```
Comment (comdw) [inline src/Conduit/AsyncPipelineManager.cs]:
```suggestion
            var options = _options.Get(name) ?? throw new ArgumentException($"Can't find named pipeline ${name}", name);
```

File: README.md [hunk 19]
```diff
@@ -1,2 +1,109 @@
...
+{
+    builder.AddPhaseFromExpression(input => input + 1);
+})
+```
+
+As part of the default pipeline builder, you have access to all of the same stuff that you would when defining a named pipeline.
+
+Once this is registered, you can access the pipeline in your code by injecting `IPipeline<T>`/`IAsyncPipeline<T>`.
+
+_**Note**: You can have different default pipelines for your synchronos and asynchronos code paths._
```
Comment (comdw) [inline README.md]:
```suggestion
_**Note**: You can have different default pipelines for your synchronous and asynchronous code paths._
```

**Review by comdw — CHANGES_REQUESTED** [review]
Sorry its mostly typos I found.  I like this new github feature where you can suggest a chance in line and the reviewer can just click to accept it as a commit! 👍

File: src/Conduit/PipelineBuilder.cs (hunk 15 above)
Comment (Im5tu) [reply src/Conduit/PipelineBuilder.cs:16]: Hmm no. Updated the check for this and the async builder

**Pull request merged**